storefront-bff/src/main/java/com/yas/storefrontbff/filter/CartAuthenticationFilter.java

```diff
@@ -0,0 +1,105 @@
+import org.springframework.http.HttpMethod;
+import org.springframework.http.server.reactive.ServerHttpRequest;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.ReactiveSecurityContextHolder;
+import org.springframework.security.core.context.SecurityContext;
+import org.springframework.security.oauth2.client.registration.ClientRegistration;
+import org.springframework.security.oauth2.client.registration.ReactiveClientRegistrationRepository;
+import org.springframework.stereotype.Component;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.reactive.function.BodyInserters;
+import org.springframework.web.reactive.function.client.WebClient;
+import org.springframework.web.server.ServerWebExchange;
+import org.springframework.web.server.WebFilter;
+import org.springframework.web.server.WebFilterChain;
+import org.springframework.web.util.UriComponentsBuilder;
+import reactor.core.publisher.Mono;
+import reactor.util.annotation.NonNull;
+
+import java.net.URI;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+@Component
+@Slf4j
+public class CartAuthenticationFilter implements WebFilter {
+    private final WebClient webClient;
+    private final ServiceUrlConfig serviceUrlConfig;
+    private final ReactiveClientRegistrationRepository clientRegistrationRepository;
+    private static final String GRANT_TYPE = "grant_type";
+
+    public CartAuthenticationFilter(WebClient webClient,
+                                    ServiceUrlConfig serviceUrlConfig,
+                                    ReactiveClientRegistrationRepository clientRegistrationRepository) {
+        this.webClient = webClient;
+        this.serviceUrlConfig = serviceUrlConfig;
+        this.clientRegistrationRepository = clientRegistrationRepository;
+    }
+
+    @Override
+    @NonNull
+    public Mono<Void> filter(ServerWebExchange exchange, WebFilterChain chain) {
+
+        ServerHttpRequest request = exchange.getRequest();
+        if (request.getPath().toString().contains("/api/cart/storefront/carts")
+                && Objects.equals(request.getMethod(), HttpMethod.POST)) {
+            AtomicBoolean isAuthenticated = new AtomicBoolean(false);
+
+            return ReactiveSecurityContextHolder.getContext()
+                    .map(SecurityContext::getAuthentication)
+                    .map(Authentication::getPrincipal)
+                    .doOnNext(principal -> isAuthenticated.set(true))
+                    .doOnError(Throwable::printStackTrace)
+                    .then(Mono.defer(() -> {
+                        if (!isAuthenticated.get()) {
+                            MultiValueMap<String, String> formData = new LinkedMultiValueMap<>();
+                            ClientRegistration clientRegistration = clientRegistrationRepository.findByRegistrationId("keycloak").block();
```

Comment:
I think it would be better if we don't block() here. Maybe doOnNext or something